src/OrbitGl/GraphTrack.cpp

```diff
@@ -0,0 +1,293 @@
+        font_color);
+  }
+
+  text_renderer.AddText(text.c_str(), text_box_position[0],
+                        text_box_position[1] + text_box_size[1] - layout_->GetTextBoxHeight() +
+                            layout_->GetTextOffset(),
+                        z, text_color, font_size, text_box_size[0]);
+}
+
+template <size_t Size>
+void GraphTrack<Size>::DrawLegend(Batcher& batcher, TextRenderer& text_renderer,
+                                  const std::array<std::string, Size>& series_names,
+                                  const Color& legend_text_color, float z) {
+  uint32_t font_size = layout_->CalculateZoomedFontSize();
+
+  const float kSpaceBetweenLegendSymbolAndText = layout_->GetRightMargin();
+  const float kSpaceBetweenLegendEntries = layout_->GetRightMargin() * 2;
+  float legend_symbol_height = GetLegendHeight() / 2.f;
+  float legend_symbol_width = legend_symbol_height;
+  float x0 = pos_[0] + layout_->GetRightMargin();
+  float y0 = pos_[1] - layout_->GetTrackTabHeight() - layout_->GetTextBoxHeight() / 2.f;
+
+  for (size_t i = 0; i < series_names.size(); ++i) {
+    batcher.AddShadedBox(Vec2(x0, y0 - legend_symbol_height / 2.f),
+                         Vec2(legend_symbol_width, legend_symbol_height), z,
+                         TimeGraph::GetColor(i));
+    x0 += legend_symbol_width + kSpaceBetweenLegendSymbolAndText;
+
+    float legend_text_width = text_renderer.GetStringWidth(series_names[i].c_str(), font_size);
+    Vec2 legend_text_box_size(legend_text_width, layout_->GetTextBoxHeight());
+    Vec2 legend_text_box_position(x0, y0 - layout_->GetTextBoxHeight() / 2.f);
+    text_renderer.AddText(series_names[i].c_str(), legend_text_box_position[0],
+                          legend_text_box_position[1] + layout_->GetTextOffset(), z,
+                          legend_text_color, font_size, legend_text_box_size[0]);
+    x0 += legend_text_width + kSpaceBetweenLegendEntries;
+  }
+}
+
+template <size_t Size>
+void GraphTrack<Size>::DrawSeries(Batcher* batcher, uint64_t min_tick, uint64_t max_tick, float z) {
+  std::array<std::string, Size> series_names = series_.names();
+  std::map<uint64_t, std::array<double, Size>> time_to_series_values = series_.value_map();
+  auto it = time_to_series_values.upper_bound(min_tick);
+  if (it == time_to_series_values.end()) return;
+  if (it != time_to_series_values.begin()) --it;
+  std::vector<float> last_normalized_values = std::vector<float>(it->second.size());
+  std::transform(
+      it->second.begin(), it->second.end(), last_normalized_values.begin(),
+      [this](double value) { return static_cast<float>((value - min_) * inv_value_range_); });
+  uint64_t previous_time = min_tick;
+  float base_y = pos_[1] - size_[1] + layout_->GetTrackBottomMargin();
+  float content_height =
+      size_[1] - layout_->GetTrackTabHeight() - layout_->GetTrackBottomMargin() - GetLegendHeight();
+
+  for (++it; it != time_to_series_values.end(); ++it) {
```

Comment:
It took me a while to understand everything that is going on here. I have an idea that you could use to make this (hopefully) simpler. 
My thought process: What needs to be done here? -> Rectangles need to be drawn for each value in the range between `min_tick` and `max_tick`. Maybe there can be a helper function in TimeSeries that returns all these times and values. Then iterate over the values and draw each. How can these helper function be implemented efficiently, so without making a copy of all the data. This requires a "view" onto the underlying data (time series values) (similar to `string_view` for `string`). This should be possible with `absl::Span`. 
So the helper function could be something like this `absl::Span<const TimeSeriesEntry<Size>> GetValuesForRange(uint64_t min, uint64_t max)`. This then hopefully makes this current `DrawSeries` function cleaner. And as a bonus, this `GetValuesForRange` can be unit tested (yay).

I am unsure if its possible how I envision this. @beckerhe do you think this is a viable and good idea?

Reply:
I like this idea! I have added the `GetEntriesInTimeRange` method to return iterators to the first and last time series entries within the time range. Also added the `DrawSingleSeriesEntry` method to make the process more clear.
Thanks for the suggestion!

Reply:
Why not the `absl::Span`?

Reply:
`std::span` only applies to continuous elements, but the `std::map` is node-based container.

Reply:
Nevermind. I looked at it in more detail and absl::Span is not an option to get a view of a std::map. Then I like your current approach best!